packages/bls-kms-local/src/agent/BlsKeyManagementSystem.ts

```diff
@@ -0,0 +1,100 @@
+import { blsSign, generateBls12381G2KeyPair } from '@mattrglobal/node-bbs-signatures'
+import Debug from 'debug'
+
+import { IKey, ManagedKeyInfo, MinimalImportableKey, TKeyType } from '@veramo/core'
+import { AbstractPrivateKeyStore, ManagedPrivateKey } from '@veramo/key-manager'
+import { KeyManagementSystem } from '@veramo/kms-local'
+import { BlsManagedKeyInfoArgs, KeyType } from '../types/IBlsKeyManager'
+
+const debug = Debug('veramo:kms:bls:local')
+
+export class BlsKeyManagementSystem extends KeyManagementSystem {
+  private readonly privateKeyStore: AbstractPrivateKeyStore
+
+  constructor(keyStore: AbstractPrivateKeyStore) {
+    super(keyStore)
+    this.privateKeyStore = keyStore
+  }
+
+  async importKey(args: Exclude<MinimalImportableKey, 'kms'>): Promise<ManagedKeyInfo> {
+    if (args.type === KeyType.Bls12381G2) {
+      if (!args.type || !args.privateKeyHex || !args.publicKeyHex) {
+        throw new Error('invalid_argument: type, publicKeyHex and privateKeyHex are required to import a key')
+      }
+      const managedKey = this.asBlsManagedKeyInfo({
+        alias: args.kid,
+        privateKeyHex: args.privateKeyHex,
+        publicKeyHex: args.publicKeyHex,
+        type: args.type,
+      })
+      await this.privateKeyStore.import({ alias: managedKey.kid, ...args })
+      debug('imported key', managedKey.type, managedKey.publicKeyHex)
+      return managedKey
+    }
+    return super.importKey(args) as Promise<ManagedKeyInfo>
+  }
+
+  async createKey({ type }: { type: TKeyType }): Promise<ManagedKeyInfo> {
+    let key: ManagedKeyInfo
+    switch (type) {
+      case KeyType.Bls12381G2: {
+        const keyPairBls12381G2 = await generateBls12381G2KeyPair()
+        key = await this.importKey({
+          kms: 'local',
+          type,
+          privateKeyHex: Buffer.from(keyPairBls12381G2.secretKey).toString('hex'),
+          publicKeyHex: Buffer.from(keyPairBls12381G2.publicKey).toString('hex'),
+        })
+        break
+      }
+      default:
+        key = await super.createKey({ type })
+    }
+
+    debug('Created key', type, key.publicKeyHex)
+
+    return key
+  }
+
+  async sign({ keyRef, algorithm, data }: { keyRef: Pick<IKey, 'kid'>; algorithm?: string; data: Uint8Array | Uint8Array[] }): Promise<string> {
+    let privateKey: ManagedPrivateKey
+    try {
+      privateKey = await this.privateKeyStore.get({ alias: keyRef.kid })
+    } catch (e) {
+      throw new Error(`key_not_found: No key entry found for kid=${keyRef.kid}`)
+    }
+
+    if (privateKey.type !== KeyType.Bls12381G2 && !Array.isArray(data)) {
+      return await super.sign({ keyRef, algorithm, data })
```

Comment:
What if data is an array and type is not Bls?

Reply:
I will keep it as a single element to be compliant with the existing implementation.

Reply:
Okay but then the change from `messages: Array.isArray(data) ? data : [data],` to `messages: [data],` should be reverted

Reply:
Yeah then you can cast it right?  The whole point is that current change will fail if you submit an array. The code is too naive to expect data not to be an array already

Reply:
There is no cast anymore and no interface changes either. Veramo does not accept Uint8Array[], but Uint8Array object and I'm adding the Uint8Arrat object to an array when passing it to the signer function. I added validation, just in case.

Reply:
Okay